google-cloud-spanner/src/test/java/com/google/cloud/spanner/it/ITDmlReturningTest.java

```diff
@@ -0,0 +1,263 @@
+    return params;
+  }
+
+  @Parameterized.Parameter(0)
+  public DialectTestParameter dialect;
+
+  private String insertDmlReturning() {
+    return String.format(INSERT_DML_RETURNING, id, id, id, id);
+  }
+
+  private String updateDmlReturning() {
+    return String.format(UPDATE_DML_RETURNING, id);
+  }
+
+  private String deleteDmlReturning() {
+    return String.format(DELETE_DML_RETURNING, id);
+  }
+
+  private String deleteDml() {
+    return String.format(DELETE_DML, id);
+  }
+
+  private DatabaseClient getClient(Dialect dialect) {
+    if (dialect == Dialect.POSTGRESQL) {
+      return postgreSQLClient;
+    }
+    return googleStandardSQLClient;
+  }
+
+  @Test
+  public void dmlReturningWithExecuteUpdate() {
+    executeUpdate(DML_COUNT, insertDmlReturning());
+    // checks for multi-stmts within a txn, therefore also verifying seqNo.
+    executeUpdate(DML_COUNT * 2, updateDmlReturning(), deleteDmlReturning());
+  }
+
+  private void executeUpdate(long expectedCount, final String... stmts) {
+    final TransactionCallable<Long> callable =
+        transaction -> {
+          long rowCount = 0;
+          for (String stmt : stmts) {
+            rowCount += transaction.executeUpdate(Statement.of(stmt));
+          }
+          return rowCount;
+        };
+    TransactionRunner runner = getClient(dialect.dialect).readWriteTransaction();
+    Long rowCount = runner.run(callable);
+    assertThat(rowCount).isEqualTo(expectedCount);
+  }
+
+  @Test
+  public void dmlReturningWithExecutePartitionedUpdate() {
+    Throwable exception =
+        assertThrows(
+            SpannerException.class,
+            () ->
+                getClient(dialect.dialect)
+                    .executePartitionedUpdate(Statement.of(updateDmlReturning())));
+    assertThat(exception.getMessage()).ignoringCase().contains("not supported");
+  }
+
+  @Test
+  public void dmlReturningWithExecuteQuery() {
+    List<Struct> rows = executeQuery(DML_COUNT, insertDmlReturning());
+    assertThat(
+            getClient(dialect.dialect)
+                .singleUse(TimestampBound.strong())
+                .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V"))
+                .getLong(0))
+        .isEqualTo(1);
+
+    // executeQuery results have keys(K) sorted.
+    // Check if keys(K) and V have expected values.
+    for (int idx = 0; idx < rows.size(); idx++) {
+      assertThat(rows.get(idx).getLong("V")).isEqualTo(idx + 1);
+      assertThat(rows.get(idx).getString("K")).isEqualTo(String.format("%d-boo%d", id, idx + 1));
+    }
+    rows = executeQuery(DML_COUNT, updateDmlReturning());
+    assertThat(
+            getClient(dialect.dialect)
+                .singleUse(TimestampBound.strong())
+                .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V"))
+                .getLong(0))
+        .isEqualTo(100);
+
+    // executeQuery results have keys(K) sorted.
```

Comment:
THEN RETURN results does not guarantee sort order. Is this order coming from somewhere else?

Reply:
I've now added sorting of rows based on key(K) inside executeQuery() method.